### ISSUE TITLE
Invalid Credentials exception from inside a docker container when using service account credentials

### ISSUE THREAD

**Issue by tamsett** [googledatalab/pydatalab#41]
Hello,

We've been using the pydatalabs modules to interact with big query in an application. When running from a local machine with my personal account this works like a charm, however when using a service account from within a docker container all commands seem to fail with:

RequestException: HTTP request failed: Invalid Credentials

However when I use gcloud.bigquery I don't seem to have this problem.

In more detail I'm creating a docker image and including the head of pydatalabs in it. I'm then building this with "pip install -e ." My code is then of the form:

import datalab.bigquery as bq

table_name = "my_project:my_dataset.my_table"
bq.Table(table_name).exists()

When run I get the following:

 File "main.py", line 75, in main
    test = bq.Table(table_name).exists()
  File "/pydatalab/datalab/bigquery/_table.py", line 189, in exists
    raise e
RequestException: HTTP request failed: Invalid Credentials

I'm exporting my service account credentials to docker via a command of the form:

docker run \
         -v my_creds.json:/etc/key_file.json \
	-e GOOGLE_APPLICATION_CREDENTIALS=/etc/key_file.json \
	my_docker_image

As far as I can tell the credentials are correct and make it into the image, also when following the above methodology both gcloud.bigquery and gcloud.pubsub work.

Thanks

Matthew

**Comment by parthea** [comment]
I have a slightly different error when I use pydatalab in a clean docker image derived from `debian:jesse`

```
root@1c7027b9b3bd:/# python
Python 2.7.9 (default, Mar  1 2015, 12:57:24) 
[GCC 4.9.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import datalab.bigquery as bq
>>> bq.Query('SELECT 1')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/dist-packages/datalab/bigquery/_query.py", line 82, in __init__
    context = datalab.context.Context.default()
  File "/usr/local/lib/python2.7/dist-packages/datalab/context/_context.py", line 92, in default
    project = _project.Projects.get_default_id(credentials)
  File "/usr/local/lib/python2.7/dist-packages/datalab/context/_project.py", line 109, in get_default_id
    projects, _ = Projects(credentials)._retrieve_projects(None, 2)
  File "/usr/local/lib/python2.7/dist-packages/datalab/context/_project.py", line 80, in _retrieve_projects
    raise e
oauth2client.client.HttpAccessTokenRefreshError: invalid_scope: Empty or missing scope not allowed.
```

My error is specific to a missing scope. I was able to resolve this issue and use pydatalab in my own docker image, but I had to modify the pydatalab source code.

In [context/_utils.py](https://github.com/googledatalab/pydatalab/blob/master/datalab/context/_utils.py), I changed line 53 from
```
    return oauth2client.client.GoogleCredentials.get_application_default()
```
to
```
    credentials = oauth2client.client.GoogleCredentials.get_application_default()
    credentials = credentials.create_scoped(['https://www.googleapis.com/auth/cloud-platform'])
    return credentials
```

Here is my docker file
```
FROM debian:jessie
RUN echo "deb-src http://ftp.us.debian.org/debian testing main" >> /etc/apt/sources.list && \
    apt-get update -y && \
    apt-get install --no-install-recommends -y -q \
        curl wget unzip git vim build-essential ca-certificates pkg-config \
        libatlas-base-dev liblapack-dev gfortran \
        libpng-dev libfreetype6-dev libxft-dev \
        rsync \
        libxml2-dev \
        python2.7 python-dev python-setuptools python-zmq openssh-client && \
    easy_install pip && \
    mkdir -p /tools

RUN mkdir -p /tools/node && \
    wget -nv https://nodejs.org/dist/v4.3.2/node-v4.3.2-linux-x64.tar.gz -O node.tar.gz && \
    tar xzf node.tar.gz -C /tools/node --strip-components=1 && \
    rm node.tar.gz

ENV PATH $PATH:/tools/node/bin:/tools/google-cloud-sdk/bin
ENV PYTHONPATH /env/python

RUN mkdir -p /tools/node && \
    wget -nv https://nodejs.org/dist/v4.3.2/node-v4.3.2-linux-x64.tar.gz -O node.tar.gz && \
    tar xzf node.tar.gz -C /tools/node --strip-components=1 && \
    rm node.tar.gz
RUN /tools/node/bin/npm install \
        ws@0.4.32 \
        http-proxy@1.13.2 \
        mkdirp@0.5.1 \
        node-uuid@1.4.7 \
        bunyan@1.7.1 \
        tcp-port-used@0.1.2 \
        node-cache@3.2.0 && \
    cd / && \
    /tools/node/bin/npm install -g forever && \
    /tools/node/bin/npm install -g typescript
RUN wget -nv https://dl.google.com/dl/cloudsdk/release/google-cloud-sdk.zip && \
    unzip -qq google-cloud-sdk.zip -d tools && \
    rm google-cloud-sdk.zip && \
    tools/google-cloud-sdk/install.sh --usage-reporting=false \
        --path-update=false --bash-completion=false \
        --disable-installation-options && \
    tools/google-cloud-sdk/bin/gcloud -q components update \
        gcloud core bq gsutil compute preview alpha beta && \
    touch /tools/google-cloud-sdk/lib/third_party/google.py

RUN pip install https://github.com/googledatalab/pydatalab/archive/master.zip
ADD datalab /usr/local/lib/python2.7/dist-packages/datalab
ADD hello.js /hello.js
ADD key.json /key.json
CMD forever hello.js
```

I was able to successfully run the sample code provided.
```
import datalab.bigquery as bq
table_name = "my_project:my_dataset.my_table"
bq.Table(table_name).exists()
```

**Comment by ojarjur** [comment]
I've hit this too, and I think @parthea is right about adding a call to `create_scoped`.

I've sent out https://github.com/googledatalab/pydatalab/pull/55 to fix it. We will probably want to expand on that fix later to support additional scopes, but at least using the cloud-platform scope should improve the situation.

**Comment by parthea** [comment]
@ojarjur Can this issue be closed? I couldn't produce the exception reported: 
```
RequestException: HTTP request failed: Invalid Credentials
```

Although, I did run into a different exception related to missing scope which was resolved by #55.

**Comment by ojarjur** [comment]
@parthea Yes, we believe the root cause is fixed in the latest releases, and enough time has passed without another report, that I think we can close this.

@tamsett Thanks so much for bringing this to our attention, and If you still see this issue with the latest Datalab release, then please let us know.
